This simple exercise is run in Jupyter Notebook.

First thing first, we import pandas and sqlite3 package to be able to run SQL query within this notebook

In [15]:
import pandas as pd
import sqlite3 as sql

Then, we connect to the database. We have a database named 'SF_salaries_database.sqlite'.

In [19]:
db = "SF_salaries_database.sqlite"
conn = sql.connect(db)

Step 1: Getting to know the table

First thing first, quite important to have the feeling of the table. Let's see how many rows and column this table has.

In [41]:
query0 = '''SELECT * FROM Salaries
            ORDER BY TotalPay DESC
'''

In [42]:
df0 = pd.read_sql_query(query0,conn)
df0

Id    EmployeeName  \
0            1  NATHANIEL FORD   
1            2    GARY JIMENEZ   
2       110532     David Shinn   
3       110533      Amy P Hart   
4        36160  Gary Altenberg   
...        ...             ...   
148649  148650   Roy I Tillery   
148650  148651    Not provided   
148651  148652    Not provided   
148652  148653    Not provided   
148653  148654       Joe Lopez   

                                              JobTitle       BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY     167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)     155966.02   
2                                       Deputy Chief 3     129150.01   
3                                    Asst Med Examiner     318835.49   
4                         Lieutenant, Fire Suppression     128808.87   
...                                                ...           ...   
148649                                       Custodian             0   
148650                                    Not provided  Not Provided   
148651                                    Not provided  Not Provided   
148652                                    Not provided  Not Provided   
148653                      Counselor, Log Cabin Ranch             0   

         OvertimePay      OtherPay      Benefits   TotalPay  TotalPayBenefits  \
0                  0     400184.25                567595.43         567595.43   
1          245131.88     137811.38                538909.28         538909.28   
2                  0     342802.63      38780.04  471952.64         510732.68   
3           10712.95      60563.54      89540.23  390111.98         479652.21   
4          220909.48      13126.31      44430.12  362844.66         407274.78   
...              ...           ...           ...        ...               ...   
148649             0             0             0       0.00              0.00   
148650  Not Provided  Not Provided  Not Provided       0.00              0.00   
148651  Not Provided  Not Provided  Not Provided       0.00              0.00   
148652  Not Provided  Not Provided  Not Provided       0.00              0.00   
148653             0       -618.13             0    -618.13           -618.13   

        Year Notes         Agency Status  
0       2011        San Francisco         
1       2011        San Francisco         
2       2014        San Francisco     PT  
3       2014        San Francisco     FT  
4       2012        San Francisco         
...      ...   ...            ...    ...  
148649  2014        San Francisco     PT  
148650  2014        San Francisco         
148651  2014        San Francisco         
148652  2014        San Francisco         
148653  2014        San Francisco     PT  

[148654 rows x 13 columns]

We got 148654 rows and 13 columns from the table.
Look at some 'strange' data in last 5 rows. We focus at rows with no value there.
Let's filter it in Step 2.

Step 2: Filtering unwanted data

Here we'd like to exclude some unwanted data which contains 'Not provided' in there.

In [48]:
query1 = '''SELECT * FROM Salaries
            WHERE EmployeeName IS NOT 'Not provided' OR JobTitle IS NOT 'Not provided'
            ORDER BY TotalPay DESC
'''

In [49]:
df1 = pd.read_sql_query(query1, conn)
df1

Id      EmployeeName  \
0            1    NATHANIEL FORD   
1            2      GARY JIMENEZ   
2       110532       David Shinn   
3       110533        Amy P Hart   
4        36160    Gary Altenberg   
...        ...               ...   
148645  148646  Carolyn A Wilson   
148646  148648    Joann Anderson   
148647  148649       Leon Walker   
148648  148650     Roy I Tillery   
148649  148654         Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                                       Deputy Chief 3  129150.01   
3                                    Asst Med Examiner  318835.49   
4                         Lieutenant, Fire Suppression  128808.87   
...                                                ...        ...   
148645                       Human Services Technician          0   
148646                     Communications Dispatcher 2          0   
148647                                       Custodian          0   
148648                                       Custodian          0   
148649                      Counselor, Log Cabin Ranch          0   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25            567595.43         567595.43  2011   
1         245131.88  137811.38            538909.28         538909.28  2011   
2              0.00  342802.63  38780.04  471952.64         510732.68  2014   
3          10712.95   60563.54  89540.23  390111.98         479652.21  2014   
4         220909.48   13126.31  44430.12  362844.66         407274.78  2012   
...             ...        ...       ...        ...               ...   ...   
148645         0.00       0.00         0       0.00              0.00  2014   
148646         0.00       0.00         0       0.00              0.00  2014   
148647         0.00       0.00         0       0.00              0.00  2014   
148648         0.00       0.00         0       0.00              0.00  2014   
148649         0.00    -618.13         0    -618.13           -618.13  2014   

       Notes         Agency Status  
0             San Francisco         
1             San Francisco         
2             San Francisco     PT  
3             San Francisco     FT  
4             San Francisco         
...      ...            ...    ...  
148645        San Francisco     PT  
148646        San Francisco     PT  
148647        San Francisco     PT  
148648        San Francisco     PT  
148649        San Francisco     PT  

[148650 rows x 13 columns]

4 rows are filtered out.
We proceed step 3 to see who is the highest earner.
The highest earner we get from TotalPayBenefits column.

We try to limit to 10 rows only.

Step 3: See top 10 of the highest earner and sort it out from the highest to the least.

In [50]:
query2 = '''SELECT * FROM Salaries
            ORDER BY TotalPayBenefits DESC
            LIMIT 10
'''

In [51]:
df2 = pd.read_sql_query(query2, conn)
df2

Id          EmployeeName  \
0       1        NATHANIEL FORD   
1       2          GARY JIMENEZ   
2  110532           David Shinn   
3  110533            Amy P Hart   
4  110534  William J Coaker Jr.   
5   72926        Gregory P Suhr   
6   72927  Joanne M Hayes-White   
7  110535        Gregory P Suhr   
8  110536  Joanne M Hayes-White   
9  110537       Ellen G Moffatt   

                                         JobTitle    BasePay  OvertimePay  \
0  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18         0.00   
1                 CAPTAIN III (POLICE DEPARTMENT)  155966.02    245131.88   
2                                  Deputy Chief 3  129150.01         0.00   
3                               Asst Med Examiner  318835.49     10712.95   
4                        Chief Investment Officer  257340.00         0.00   
5                                 Chief of Police  319275.01         0.00   
6                          Chief, Fire Department  313686.01         0.00   
7                                 Chief of Police  307450.04         0.00   
8                          Chief, Fire Department  302068.00         0.00   
9                               Asst Med Examiner  270222.04      6009.22   

    OtherPay  Benefits   TotalPay  TotalPayBenefits  Year Notes  \
0  400184.25            567595.43         567595.43  2011         
1  137811.38            538909.28         538909.28  2011         
2  342802.63  38780.04  471952.64         510732.68  2014         
3   60563.54  89540.23  390111.98         479652.21  2014         
4   82313.70  96570.66  339653.70         436224.36  2014         
5   20007.06  86533.21  339282.07         425815.28  2013         
6   23236.00  85431.39  336922.01         422353.40  2013         
7   19266.72  91302.46  326716.76         418019.22  2014         
8   24165.44  91201.66  326233.44         417435.10  2014         
9   67956.20  71580.48  344187.46         415767.94  2014         

          Agency Status  
0  San Francisco         
1  San Francisco         
2  San Francisco     PT  
3  San Francisco     FT  
4  San Francisco     PT  
5  San Francisco         
6  San Francisco         
7  San Francisco     FT  
8  San Francisco     FT  
9  San Francisco     FT

We see that Nathaniel Ford and Gary Jimenez are among the top earner in San Fransisco.
But their records are in the same year.

Let's who are the top earner each year in Step 4.

Step 4: Seek out top earner employee each year

In [55]:
query3 = '''SELECT Year, EmployeeName, JobTitle, MAX(TotalPayBenefits) FROM Salaries
            WHERE EmployeeName IS NOT 'Not provided' OR JobTitle IS NOT 'Not provided' OR TotalPayBenefits > 0
            GROUP BY Year
            LIMIT 10
'''

In [56]:
df3 = pd.read_sql_query(query3, conn)
df3

Year    EmployeeName                                        JobTitle  \
0  2011  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1  2012  Gary Altenberg                    Lieutenant, Fire Suppression   
2  2013  Gregory P Suhr                                 Chief of Police   
3  2014     David Shinn                                  Deputy Chief 3   

   MAX(TotalPayBenefits)  
0              567595.43  
1              407274.78  
2              425815.28  
3              510732.68

Now, we can see who are the top earner for each year, from 2011 to 2014
However, we don't see any engineer job title in those list. I'm curious how much engineers in this dataset.

Step 5. Find Total Earning for Engineer Job Title

In [58]:
query4 = '''SELECT EmployeeName, JobTitle, TotalPayBenefits, Year FROM Salaries
            WHERE JobTitle LIKE 'Engine%' AND EmployeeName IS NOT 'Not provided'
            ORDER BY TotalPayBenefits DESC
            LIMIT 10
'''

In [59]:
df4 = pd.read_sql_query(query4, conn)
df4

EmployeeName                      JobTitle  TotalPayBenefits  Year
0     John O Funghi  Engineer/Architect Principal         258022.86  2014
1     John O Funghi  Engineer/Architect Principal         254521.86  2013
2   Reuben A Halili                      Engineer         250398.39  2014
3      Hanson W Tom  Engineer/Architect Principal         249483.69  2014
4      Hanson W Tom  Engineer/Architect Principal         247836.55  2013
5     John F Thomas                      Engineer         244675.25  2014
6     Howard H Fung                      Engineer         244675.16  2014
7  Shahnam Farhangi  Engineer/Architect Principal         243652.62  2014
8   Fariba Mahmoudi  Engineer/Architect Principal         243652.59  2014
9     John F Thomas                      Engineer         242898.03  2013

We find that top 10 engineers are earning between 240000 to 250000 anually.
Finally, in this last step, let's create a query that seek out what are the job titles in San Fransisco with top average total earning.

Step 6:
In this step, we use window function in order to see the rank for the job title sorted from the highest max total pay.

In [62]:
query5 = '''SELECT RANK() OVER(ORDER BY MAX(TotalPayBenefits) DESC) AS 'Rank by Max Total Pay',            
            JobTitle, ROUND(AVG(TotalPayBenefits),1) AS 'Average Total Pay', Year FROM Salaries
            WHERE EmployeeName IS NOT 'Not provided' OR JobTitle IS NOT 'Not provided'
            GROUP BY JobTitle
            HAVING AVG(TotalPayBenefits) > 350000
            ORDER BY 3 DESC
'''

In [63]:
df5 = pd.read_sql(query5, conn)
df5

Rank by Max Total Pay                                        JobTitle  \
0                      2                        Chief Investment Officer   
1                      3                                 Chief of Police   
2                      4                          Chief, Fire Department   
3                      1  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
4                      6                      Gen Mgr, Public Trnsp Dept   
5                      5                    Dep Dir for Investments, Ret   
6                      7                                           Mayor   

   Average Total Pay  Year  
0           436224.4  2014  
1           411732.3  2013  
2           408865.3  2013  
3           399211.3  2011  
4           380696.4  2013  
5           355731.3  2013  
6           354212.9  2014

We got General Manager-Metropolitan Transit Authority ranks 1 in total pay but lies in no 4 in average total pay. The top average total pay is Chief Investment Office with 436224.4 annual earning.